contracts/src/v0.8/automation/upkeeps/UpkeepBalanceMonitor.sol

```diff
@@ -0,0 +1,232 @@
+        count++;
+        availableFunds -= topUpAmount;
+      }
+      if (count == config.maxBatchSize) {
+        break;
+      }
+    }
+    if (count < numUpkeeps) {
+      assembly {
+        mstore(needsFunding, count)
+        mstore(topUpAmounts, count)
+      }
+    }
+    return (needsFunding, topUpAmounts);
+  }
+
+  /// @notice Called by the keeper/owner to send funds to underfunded upkeeps
+  /// @param upkeepIDs the list of upkeep ids to fund
+  /// @param topUpAmounts the list of amounts to fund each upkeep with
+  /// @dev We explicitly choose not to verify that input upkeepIDs are included in the watchlist. We also
+  /// explicity permit any amount to be sent via topUpAmounts; it does not have to meet the criteria
+  /// specified in getUnderfundedUpkeeps(). Here, we are relying on the security of automation's OCR to
+  /// secure the output of getUnderfundedUpkeeps() as the input to topUp(), and we are treating the owner
+  /// as a privileged user that can perform arbitrary top-ups to any upkeepID.
+  function topUp(uint256[] memory upkeepIDs, uint96[] memory topUpAmounts) public {
+    IAutomationForwarder forwarder = s_forwarder;
+    if (msg.sender != address(s_forwarder) && msg.sender != owner()) revert OnlyForwarderOrOwner();
+    if (upkeepIDs.length != topUpAmounts.length) revert InvalidTopUpData();
+    address registryAddress = address(forwarder.getRegistry());
+    for (uint256 i = 0; i < upkeepIDs.length; i++) {
+      try LINK_TOKEN.transferAndCall(registryAddress, topUpAmounts[i], abi.encode(upkeepIDs[i])) returns (
+        bool success
+      ) {
+        if (success) {
+          emit TopUpSucceeded(upkeepIDs[i], topUpAmounts[i]);
+          continue;
+        }
+      } catch {}
+      emit TopUpFailed(upkeepIDs[i]);
+    }
+  }
+
+  // ================================================================
+  // |                    AUTOMATION COMPATIBLE                     |
+  // ================================================================
+
+  /// @notice Gets list of upkeeps ids that are underfunded and returns a keeper-compatible payload.
+  /// @return upkeepNeeded signals if upkeep is needed, performData is an abi encoded list of subscription ids that need funds
+  function checkUpkeep(
+    bytes calldata
+  ) external view whenNotPaused returns (bool upkeepNeeded, bytes memory performData) {
+    (uint256[] memory needsFunding, uint256[] memory topUpAmounts) = getUnderfundedUpkeeps();
+    upkeepNeeded = needsFunding.length > 0;
+    if (upkeepNeeded) {
+      performData = abi.encode(needsFunding, topUpAmounts);
+    }
+    return (upkeepNeeded, performData);
+  }
+
+  /// @notice Called by the keeper to send funds to underfunded addresses.
+  /// @param performData the abi encoded list of addresses to fund
+  function performUpkeep(bytes calldata performData) external whenNotPaused {
+    (uint256[] memory upkeepIDs, uint96[] memory topUpAmounts) = abi.decode(performData, (uint256[], uint96[]));
```

Comment:
should the `uint256[], uint96[]` align with `uint256[] memory needsFunding, uint256[] memory topUpAmounts` as in L137? Since you tested this, I guess it doesn't matter? But curious to know the reasoning behind those types :)